plugin/train.py

```diff
@@ -0,0 +1,136 @@
+
+
+def gluoncv_train(config_dict, model_params):
+    """GluonCV Training Script
+
+    arg str config_path: default dictionary containing relevant configuration
+                         files, including classes and training_data_dir.
+    arg str model_params: path to pretrained model
+
+    return: trained model
+    """
+
+    epochs = 3
+    per_device_batch_size = 32
+
+    lr_decay = 0.1
+    lr_decay_epoch = [10, 20, 30, 40, np.inf]
+
+    num_cpus = 1
+    ctx = [mx.cpu(i) for i in range(num_cpus)]
+    batch_size = per_device_batch_size * max(num_cpus, 1)
+
+    optimizer = 'nag'
+
+    optimizer_params = {'learning_rate': 0.01, 'wd': 0.0001, 'momentum': 0.9}
+
+    model = model_build(config_dict['model']['nb_classes'], model_params)
+    model.initialize(mx.init.Xavier(), ctx=ctx)
+
+    trainer = gluon.Trainer(model.collect_params(),
+                            optimizer, optimizer_params)
+
+    train_path = config_dict['trainer']['options']['training_data_dir']
+    val_path = config_dict['trainer']['options']['validation_data_dir']
+
+    train_data = gluon.data.DataLoader(
+        gluon.data.vision.ImageFolderDataset(train_path)
+                         .transform_first(make_transform()),
+        batch_size=batch_size, shuffle=True, last_batch='keep')
+    val_data = gluon.data.DataLoader(
+        gluon.data.vision.ImageFolderDataset(val_path)
+                         .transform_first(make_transform()),
+        batch_size=batch_size, shuffle=False, last_batch='keep')
+
+    train_metric = mx.metric.F1()
+    loss_fn = gluon.loss.SoftmaxCrossEntropyLoss()
+    num_batch = len(train_data)
+    lr_decay_count = 0
+
+    for epoch in range(epochs):
+        tic = time.time()
+        train_loss = 0
+        train_metric.reset()
+
+        if epoch == lr_decay_epoch[lr_decay_count]:
+            trainer.set_learning_rate(trainer.learning_rate*lr_decay)
+            lr_decay_count += 1
+
+        for batch in train_data:
+            data = gluon.utils.split_and_load(batch[0], ctx_list=ctx,
+                                              batch_axis=0)
+            label = gluon.utils.split_and_load(batch[1], ctx_list=ctx,
+                                               batch_axis=0)
+            with ag.record():
+                outputs = [model(X) for X in data]
+                loss = [loss_fn(yhat, y) for yhat, y in zip(outputs, label)]
+            for l in loss:
+                l.backward()
+            trainer.step(batch_size)
+            train_loss += sum([l.mean().asscalar() for l in loss]) / len(loss)
+
+            train_metric.update(label, outputs)
+
+        _, train_f1 = train_metric.get()
+        train_loss /= num_batch
+
+        _, val_f1 = test(model, ctx, val_data)
+
+        print('[Epoch %d] Train-f1: %.3f, loss: %.3f |'
+              'Val-f1: %.3f | time: %.1f' %
+              (epoch, train_f1, train_loss, val_f1, time.time() - tic))
+
+        model.save_parameters(model_params)
```

Comment:
The PyDocs says that `model_params` is the path to the pretrained model weights, but that's not how you're using that variable here. A better name would be `model_path` and then just update the comment.

Reply:
You're only saving the model at the very end, so that means the periodic sync has nothing to upload. Getting checkpointing to work is going to take some time that we don't have right now, so I would just add a TODO comment about that.